File: 240520/자동 잔디 깎이/automatic-lawn-mowing.py
n = int(input())
move = [tuple(input().split()) for _ in range(n)]

indicator = {'E': (0, 1), 'W': (0, -1), 'S': (1, 0), 'N': (-1, 0)}
x = 0
y = 0
visit_map = {} # {(x,y): [step, min_visit]}
step = 0

for i in range(n):
    direction, distance = move[i]

    for j in range(int(distance)):
        step += 1
        x += indicator[direction][1]
        y -= indicator[direction][0]

        if (x,y) in visit_map:
            min_cnt = step - visit_map[(x,y)][0]
            visit_map[(x,y)][1] = min(visit_map[(x,y)][1], min_cnt)
            visit_map[(x,y)][0] = step
        else:
            visit_map[(x,y)] = [step, 10 * 100]

min_visit = 10 * 100
for x, y in visit_map:
    m_visit = visit_map[(x,y)][1]
    min_visit = min(min_visit, m_visit)
        
print(min_visit if min_visit < 10*100 else -1)

File: 240520/자동 잔디 깎이/automatic-lawn-mowing.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int n = Integer.parseInt(scan.nextLine());
        String[][] move = new String[n][2];
        for(int i = 0; i < n; i++) {
            move[i] = scan.nextLine().split(" ");
        }
        
        int maxLen = 10 * 100;
        Map<String, Integer[]> indicator = new HashMap<>();
        indicator.put("E", new Integer[]{0, 1});
        indicator.put("W", new Integer[]{0, -1});
        indicator.put("S", new Integer[]{1, 0});
        indicator.put("N", new Integer[]{-1, 0});
        
        int[][] visit = new int[maxLen*2][maxLen*2];
        int x = maxLen, y = maxLen, minVisit = maxLen;
        int step = 0;
        
        for(int i=0; i < n; i++) {
            String direction = move[i][0];
            int distance = Integer.parseInt(move[i][1]);
            for(int j=0; j < distance; j++) {
                step++;
                x += indicator.get(direction)[0];
                y += indicator.get(direction)[1];

                if(visit[x][y] < maxLen && visit[x][y] != 0) {
                    minVisit = Math.min(minVisit, step - visit[x][y]);
                } else {
                    visit[x][y] = step;
                }
            }
        }

        System.out.println(minVisit == maxLen ? -1 : minVisit);
    }
}